indexer/src/main.ts

```diff
@@ -73,206 +79,238 @@ export default async function transform({
-      store.push({
-        collection: "receipts",
-        data: {
-          receipt: ethReceipt,
-        },
-      });
-    }
-  });
+  const processedTransactions = processTransactions(
+    transactions,
+    blockInfo,
+    cumulativeGasUsages,
+  );
+  updateStoreWithTransactions(store, processedTransactions);
 
   const ethHeader = await toEthHeader({
     header: header,
     gasUsed: cumulativeGasUsed,
     logsBloom: blockLogsBloom,
     receiptRoot: receiptTrie.root(),
     transactionRoot: transactionTrie.root(),
-    blockNumber,
-    blockHash,
-    isPendingBlock,
+    ...blockInfo,
   });
   store.push({
-    collection: "headers",
+    collection: Collection.Headers,
     data: { header: ethHeader },
   });
 
   return store;
 }
+
+function createBlockInfo(header: BlockHeader): BlockInfo {
+  const blockNumber = padString(toHexString(header.blockNumber), 8);
+  const blockHash = padString(header.blockHash, 32);
+  const isPendingBlock = blockHash === NULL_BLOCK_HASH;
+  return { blockNumber, blockHash, isPendingBlock };
+}
+
+function processEvent(blockInfo: BlockInfo) {
+  return (event: EventWithTransaction): ProcessedEvent | null => {
+    const typedEthTx = toTypedEthTx({ transaction: event.transaction });
+    // Can be null if:
+    // 1. The transaction is missing calldata.
+    // 2. The transaction is a multi-call.
+    // 3. The length of the signature array is different from 5.
+    // 4. The chain id is not encoded in the v param of the signature for a
+    //    Legacy transaction.
+    // 5. The deserialization of the transaction fails.
+    if (!typedEthTx) return null;
+
+    const ethTx = typedTransactionToEthTx({
+      typedTransaction: typedEthTx!,
+      receipt: event.receipt,
+      ...blockInfo,
+    });
+    // Can be null if:
+    // The typed transaction is missing a signature param (v, r, s).
+    if (!ethTx) return null;
+
+    const ethLogs = event.receipt.events
+      .map((e) => toEthLog({ transaction: ethTx, event: e, ...blockInfo }))
+      // Can be null if:
+      // 1. The event is part of the defined ignored events (see IGNORED_KEYS).
+      // 2. The event has an invalid number of keys.
+      .filter((e): e is JsonRpcLog => e !== null);
+
+    const ethLogsIndexed = ethLogs.map((log, index) => ({
+      ...log,
+      logIndex: index.toString(),
+    }));
+
+    const ethReceipt = toEthReceipt({
+      transaction: ethTx as JsonRpcTx,
+      logs: ethLogsIndexed,
+      event: event.event,
+      cumulativeGasUsed: 0n, // This will be updated later
+      ...blockInfo,
+    });
+
+    return { event, typedEthTx, ethTx, ethLogs: ethLogsIndexed, ethReceipt };
+  };
+}
+
+function accumulateGasAndUpdateStore(
+  processedEvents: ProcessedEvent[],
+  store: Array<StoreItem>,
+  blockLogsBloom: Bloom,
+): { cumulativeGasUsed: bigint; cumulativeGasUsages: bigint[] } {
+  let cumulativeGasUsed = 0n;
+  const cumulativeGasUsages: bigint[] = [];
+
+  if (!Array.isArray(processedEvents)) {
```

Comment:
how can `processedEvents` not be an array since the type is `ProcessedEvent[]` ?

Reply:
fixed, thanks